driver/src/solution_submission.rs

```diff
@@ -103,33 +103,57 @@ impl<'a> StableXSolutionSubmitter<'a> {
             gas_price_estimating,
         }
     }
+
+    /// Turn a method error from a solution submission into a SolutionSubmissionError.
+    async fn make_error(
+        &self,
+        batch_index: U256,
+        solution: Solution,
+        err: MethodError,
+    ) -> SolutionSubmissionError {
+        if let Some(tx) = extract_transaction_receipt(&err) {
+            if let Some(block_number) = tx.block_number {
+                if let Err(err) = self
+                    .contract
+                    .get_solution_objective_value(batch_index, solution, Some(block_number.into()))
+                    .await
+                {
+                    return SolutionSubmissionError::from(err);
+                }
+            }
+        }
+        return SolutionSubmissionError::Unexpected(err.into());
+    }
 }
 
 impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
     fn get_solution_objective_value(
         &self,
         batch_index: U256,
         solution: Solution,
-    ) -> Result<U256, SolutionSubmissionError> {
-        // NOTE: Compare with `>=` as the exchange's current batch index is the
-        //   one accepting orders and does not yet accept solutions.
-        while batch_index.as_u32() >= self.contract.get_current_auction_index().wait()? {
-            info!("Solved batch is not yet accepting solutions, waiting for next batch.");
-            thread::sleep(POLL_TIMEOUT);
-        }
+    ) -> BoxFuture<Result<U256, SolutionSubmissionError>> {
+        async move {
+            // NOTE: Compare with `>=` as the exchange's current batch index is the
+            //   one accepting orders and does not yet accept solutions.
+            while batch_index.as_u32() >= self.contract.get_current_auction_index().await? {
+                info!("Solved batch is not yet accepting solutions, waiting for next batch.");
+                thread::sleep(POLL_TIMEOUT);
```

Comment:
Should we use https://docs.rs/async-std/1.1.0/async_std/task/fn.sleep.html here to not block on the sleep?

Reply:
Eventually we need to decide on a runtime (probably asyncstd or tokio) and then we can use their sleep function. I was hoping to do that after this PR. In general it is wrong to do blocking IO or sleeping like this in an async function. In this case it is still fine because we're not yet running the future on any executor.
But this is a pitfall so I should have mentioned this somewhere.

Edit: I *thought* we couldn't use this function if the future isn't running on async-std's executor because that is what I remembered from tokio, but maybe I'm wrong? I'll have to check this.

Reply:
I decided to use the https://docs.rs/futures-timer/3.0.2/futures_timer/ crate for now. It is executor agnostic and internally used by async-std.